ai_diffusion/ui/widget.py

```diff
@@ -590,25 +593,75 @@ def move_cursor_to_end(self):
             self._single.setCursorPosition(len(self._single.text()))
 
 
+# SpinBox variant that allows manually entering strength values,
+# but snaps to model_steps on step actions (scrolling, arrows, arrow keys).
+class StrengthSpinBox(QSpinBox):
+    min_steps: int | None
+    max_steps: int | None
+
+    def __init__(self, parent=None):
+        super().__init__(parent)
+        # safe default
+        self.min_steps = None
+        self.max_steps = None
+        # for manual input
+        self.setMinimum(1)
+        self.setMaximum(100)
+
+    def step_info(self, value: int | None = None, pretty: bool = False) -> tuple[int, int] | None:
+        if self.min_steps is None or self.max_steps is None:
+            return None
+        # See workflow.py:_apply_strength
+        strength = (value or self.value()) / 100
+        steps = self.max_steps * strength
+        # round is more accurate, but int ensures that round steps end up on round values
```

Comment:
I don't really understand this comment. Why is this needed? Shouldn't it always follow workflow computation (round)?

Reply:
The problem is: say you have ten steps. Thanks to round(), you map 45-55 to 10/20. But snapping to "45% (10/20)" would be very strange. And finding the midpoint manually would make the code awkward. So `pretty=True` changes the interval to *start* at the natural midpoint.

Reply:
Okay I see what you mean. It's still super confusing if you just look at the `apply_strength` code, it says `round_up` but actually does a `ceil`. And it technically gives wrong results if you don't use it in this exact searching-the-next-threshold iterative way.

Why not loop until you find the next threshold using the original function with round. Say at 45 steps goes from 9 to 10. Then return `steps / max_steps`, which is 50.

Reply:
Should be done now.